src/Optimization/hiopFRProb.cpp

```diff
@@ -0,0 +1,852 @@
+  wrk_c_->axpy(-1.0, *wrk_eq_);
+  wrk_c_->axpy(1.0, *wrk_eq_);
+
+  wrk_ineq_->copy_from_starting_at(x, pi_st_, m_ineq_); //pi
+  wrk_ineq_->copy_from_starting_at(x, ni_st_, m_ineq_); //ni
+  wrk_d_->axpy(-1.0, *wrk_ineq_);
+  wrk_d_->axpy(1.0, *wrk_ineq_);
+
+  // assemble the full vector
+  wrk_c_->copyToStarting(*wrk_dual_, 0);
+  wrk_d_->copyToStarting(*wrk_dual_, m_eq_);
+
+  wrk_dual_->copyTo(cons);
+
+  return true;
+}
+
+bool hiopFRProbSparse::eval_Jac_cons(const long long& n, const long long& m,
+                                          const long long& num_cons,
+                                          const long long* idx_cons,
+                                          const double* x,
+                                          bool new_x,
+                                          const int& nnzJacS,
+                                          int* iJacS,
+                                          int* jJacS,
+                                          double* MJacS)
+{
+  return false;
+}
+
+/// @pre assuming Jac of the original prob is sorted
+bool hiopFRProbSparse::eval_Jac_cons(const long long& n,
+                                          const long long& m,
+                                          const double* x,
+                                          bool new_x,
+                                          const int& nnzJacS,
+                                          int* iJacS,
+                                          int* jJacS,
+                                          double* MJacS)
+{
+  assert( n == n_);
+  assert( m == m_);
+
+  assert(nnzJacS == nlp_base_->get_nnz_Jaceq() + nlp_base_->get_nnz_Jacineq() + 2 * (m_));
+
+  hiopMatrixSparse* Jac_c = dynamic_cast<hiopMatrixSparse*>(solver_base_.get_Jac_c());
+  hiopMatrixSparse* Jac_d = dynamic_cast<hiopMatrixSparse*>(solver_base_.get_Jac_d());
+
+  // shortcut to the original Jac
+  int *irow_c = Jac_c->i_row();
+  int *jcol_c = Jac_c->j_col();
+  int *irow_d = Jac_d->i_row();
+  int *jcol_d = Jac_d->j_col();
+
+  // assuming Jac is sorted!
+  int nnz_Jac_c_base = nlp_base_->get_nnz_Jaceq();
+  int nnz_Jac_d_base = nlp_base_->get_nnz_Jacineq();
+
+  int k;
+  int k_base;
+  int last_row_idx;
+
+  // extend Jac to the p and n parts
+  if(iJacS != NULL && jJacS != NULL){
+    // Jac for d(x) - p + n
+    last_row_idx = m_;
+    k_base = nnz_Jac_d_base - 1;
+    for(k = nnz_Jac_c_ + nnz_Jac_d_ - 1; k >= nnz_Jac_c_ && k_base >= 0; ) {
+      int row_in_jac_d = irow_d[k_base];
+      int row_idx = row_in_jac_d + m_eq_;
+      if(row_idx != last_row_idx) {
+        // n
+        iJacS[k] = row_idx;
+        jJacS[k] = ni_st_ + row_in_jac_d;
+        // p
+        iJacS[k-1] = row_idx;
+        jJacS[k-1] = pi_st_ + row_in_jac_d;
+
+        last_row_idx = row_idx;
+        k = k-2;
+      } else {
+        // x
+        iJacS[k] = row_idx;
+        jJacS[k] = jcol_d[k_base];
+        k--;
+        k_base--;
+      }
+    }
```

Comment:
This does not seem to be portable nor a thread-safe implementation. Is it expected this function will run sequentially and on CPU only?

Reply:
There are several instances like this in `hiopFRProbSparse` class where member method access raw data. It would certainly be worthwhile to encapsulate these computations in specialized matrix kernels and keep `hiopFRProbSparse` implementation generic.

Reply:
Good point, and I guess I know how to do it.
It seems that we cannot test the corresponding RAJA implementation due to the lack of `gpu` mode for sparse instance. Therefore probably I will create a new ticket to address this issue.
Will discuss with @cnpetra tomorrow.

Reply:
as per our discussion, to prioritize the merging of this, on which other things depend, @nychiang please fill an issue with this and we'll follow up with a matrix method and associated device implementation for this.